mlrun/serving/merger.py

```diff
@@ -0,0 +1,154 @@
+
+        self._uplinks = None
+        self._cache = {}
+        self._sequence = 0
+        self._get_join_key = None
+        self._queue_len = max_behind or 64  # default queue is 64 entries
+        self._keys_queue = []
+
+    def post_init(self, mode="sync"):
+        # auto detect number of uplinks or use user specified value
+        self._uplinks = self.expected_num_events or (
+            len(self._graph_step.after) if self._graph_step else 0
+        )
+
+        # function to extract the join key from the event
+        key_path = self.key_path or "event.id"
+        self._get_join_key = eval("lambda event: " + key_path, {}, {})
+        self._keys_queue = [None] * self._queue_len
+
+    def get_join_key(self, event):
+        """function extract the join key from the event, can be overwritten by sub class"""
+        return self._get_join_key(event)
+
+    async def _do(self, event):
+        if event is storey.dtypes._termination_obj:
+            return await self._do_downstream(storey.dtypes._termination_obj)
+        else:
+            element = self._get_event_or_body(event)
+            fn_result = self._merge_events(element)
+            if fn_result:
+                mapped_event = self._user_fn_output_to_event(event, fn_result)
+                await self._do_downstream(mapped_event)
+
+    def _merge_events(self, event):
+        # skip if only one uplink
+        if self._uplinks <= 1:
+            return event
+
+        merge_key = self.get_join_key(event)
+        if merge_key in self._cache:
+            # old events with that key already exist (cached)
+            entry: CacheEntry = self._cache[merge_key]
+            if entry.is_late:
+                self.context.logger.info(f"event id {merge_key} arrived late")
+                return None
+
+            if self.context.verbose:
+                self.context.logger.info(
+                    f"event id {merge_key} part {entry.arrived} arrived"
+                )
+            entry.arrived += 1
+            entry.events.append(event)
+
+            if entry.arrived == self._uplinks:
+                # expected number of events arrived, can merge them
+                if self.context.verbose:
+                    self.context.logger.info(
+                        f"event {merge_key}, all {entry.arrived} parts arrived"
+                    )
+                del self._cache[merge_key]
+                self._keys_queue[entry.sequence % self._queue_len] = None
+                return self.merge_function(event, entry.events)
+        else:
+            # first time the key arrives
+            self._cache[merge_key] = CacheEntry(self._sequence, event)
+            queue_index = self._sequence % self._queue_len
+            old_key = self._keys_queue[queue_index]
+
+            if self.context.verbose:
+                self.context.logger.info(f"new event id {merge_key} arrived")
+
+            if old_key is not None:
```

Comment:
Note that this assumes that events are processed in order and if original event A follows B, then all parts of A arrive before all parts of B. This would not be the case if some parts get dropped due to errors, filters etc.
The result of non-order may be that the queue cell for the current sequence is occupied, but there is a cell with higher sequence number which is free.

Reply:
thats by design, we allow X number of out of order events, other mechanisms will be more expensive